Index: Menu.cpp
===================================================================
--- Menu.cpp	(revision 22)
+++ Menu.cpp	(working copy)
@@ -465,9 +465,16 @@
 void OnContextMenu(WindowInfo* win, int x, int y)
 {
     assert(win->IsDocLoaded());
-    if (!win->IsDocLoaded())
+	if (!win->IsDocLoaded())
         return;
 
+	BetsyNetPDFUnmanagedApi* betsyApi = (BetsyNetPDFUnmanagedApi*)win->betsyApi;
+	if(betsyApi)
+	{
+		betsyApi->CheckOnRequestContextMenu(win, x, y);
+		return;
+	}
+
     PageElement *pageEl = win->dm->GetElementAtPos(PointI(x, y));
     ScopedMem<WCHAR> value(pageEl ? pageEl->GetValue() : NULL);
     CrashIf(value && !pageEl);
