Index: src/WindowInfo.cpp
===================================================================
--- src/WindowInfo.cpp	(revision 22)
+++ src/WindowInfo.cpp	(working copy)
@@ -39,6 +39,8 @@
 {
     ZeroMemory(&selectionRect, sizeof(selectionRect));
 
+	betsyApi = NULL;
+
     dpi = win::GetHwndDpi(hwndFrame, &uiDPIFactor);
     touchState.panStarted = false;
     buffer = new DoubleBuffer(hwndCanvas, canvasRc);
Index: src/WindowInfo.h
===================================================================
--- src/WindowInfo.h	(revision 22)
+++ src/WindowInfo.h	(working copy)
@@ -57,6 +57,7 @@
 public:
     WindowInfo(HWND hwnd);
     ~WindowInfo();
+	void* betsyApi;
 
     // TODO: error windows currently have
     //       !IsAboutWindow() && !IsDocLoaded()
